File: jvm/core/main/scala/com/webank/eggroll/core/deepspeed/store/meta/RendezvousStoreAddRequest.scala
package com.webank.eggroll.core.deepspeed.store.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.meta.Deepspeed

import scala.language.implicitConversions

case class RendezvousStoreAddRequest(prefix: String, key: K, amount: Long)

object RendezvousStoreAddRequest {
  implicit def serialize(src: RendezvousStoreAddRequest): Array[Byte] = {
    val builder = Deepspeed.StoreAddRequest.newBuilder()
      .setPrefix(src.prefix)
      .setKey(ByteString.copyFrom(src.key.toArray))
      .setAmount(src.amount)
    builder.build().toByteArray
  }

  implicit def deserialize(byteString: ByteString): RendezvousStoreAddRequest = {
    val src = Deepspeed.StoreAddRequest.parseFrom(byteString)
    RendezvousStoreAddRequest(
      prefix = src.getPrefix,
      key = src.getKey.toByteArray.toVector,
      amount = src.getAmount
    )
  }
}

File: jvm/new_core/src/main/java/com/eggroll/core/deepspeed/store/RendezvousStoreAddRequest.java
package com.eggroll.core.deepspeed.store;

import com.eggroll.core.utils.JsonUtil;
import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;
import com.webank.eggroll.core.meta.Deepspeed;
import lombok.Data;

import java.util.Vector;

@Data
public class RendezvousStoreAddRequest<K> {
    private String prefix;
    private K key;
    private long amount;

    public RendezvousStoreAddRequest(String prefix, K key, long amount) {
        this.prefix = prefix;
        this.key = key;
        this.amount = amount;
    }

    public static <K> byte[] serialize(RendezvousStoreAddRequest<K> src) {
        Deepspeed.StoreAddRequest.Builder builder = Deepspeed.StoreAddRequest.newBuilder()
                .setPrefix(src.getPrefix())
                .setKey(ByteString.copyFrom(JsonUtil.convertToByteArray(src.getKey())))
                .setAmount(src.getAmount());
        return builder.build().toByteArray();
    }

    public static <K> RendezvousStoreAddRequest<K> deserialize(ByteString byteString) throws InvalidProtocolBufferException {
        Deepspeed.StoreAddRequest proto = Deepspeed.StoreAddRequest.parseFrom(byteString);
        String prefix = proto.getPrefix();
        Vector<Byte> key = byteArrayToVector(proto.getKey().toByteArray());
        long amount = proto.getAmount();
        return new RendezvousStoreAddRequest(prefix, key, amount);
    }

    private static Vector<Byte> byteArrayToVector(byte[] byteArray) {
        Vector<Byte> vector = new Vector<>();
        for (byte b : byteArray) {
            vector.add(b);
        }
        return vector;
    }
}